Plugin DSL

oguzkocer opened the pull request:
This PR updates the project to use Gradle plugin DSL as part of our ongoing effort to update all our projects and also closes #15703 as outlined [in this comment](https://github.com/wordpress-mobile/WordPress-Android/issues/15703#issuecomment-1000936450). This was a bit trickier than most projects, but it was not unexpected due to its size. @wzieba would you be up to reviewing this PR? You have had experience with this transition in multiple projects, so I think you're the most likely candidate to catch anything missing or wrong. As both of us experienced, there are some unexpected twists with the plugin DSL, so I am not fully confident in all the changes. (also feel free to add your thoughts in the discussion in #15703)

I tried to keep this transition as 1 to 1 as I can, but there are still a few things worth mentioning:
* I did not port `checkstyle` to the plugin DSL. I tried at first, but it requires some extra research so I left it for a separate PR.
* I had to address the issue from #15703 where Kotlin dependencies were being resolved to `1.6.0` which happened in 320eed4952cbcddcfc76a36cce445a77e3ad6b75 (which fixed a merge conflict issue). As far as I could tell, there were no dependencies that required Kotlin `1.6.0`, but I might have missed a dynamic dependency. In any case, I think we should start forcing Gradle to resolve dependencies to the version we specify as in 9bdab5230e5b87693d666e0a7d9f037cdb7bc544 and this was a good example why.

**To test:**
* Smoke test the app

## Regression Notes
1. Potential unintended areas of impact
Incorrect dependency version resolutions

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

oguzkocer committed:
Moves most root plugins to plugin dsl

oguzkocer committed:
Move fetchstyle plugin to plugin dsl

oguzkocer committed:
Most plugins in subprojects converted to plugin DSL

oguzkocer committed:
Remove checkstyle plugin from libs/editor since it's applied for all projects from root

oguzkocer committed:
Use plugin DSL for checkstyle and detekt plugins

oguzkocer committed:
Move androidx.navigation.safeargs.kotlin to plugin dsl

oguzkocer committed:
Remove buildscript section from root build file

oguzkocer committed:
Move google-services plugin to plugin DSL

oguzkocer committed:
Merge branch 'improve/build-gradle-file-format' into plugin-dsl

oguzkocer committed:
Use old syntax for applying checkstyle plugin

oguzkocer committed:
Adds AGP to root build.gradle without applying

oguzkocer committed:
Update kotlinVersion to 1.5.32

oguzkocer committed:
Force kotlin dependency resolution to specified value

wzieba commented on WordPress/build.gradle:
```diff
@@ -284,7 +275,7 @@ dependencies {
     testImplementation "$gradle.ext.storiesAndroidPhotoEditorPath:$storiesVersion"
     implementation project(path:':libs:image-editor::ImageEditor')
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlinVersion"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$gradle.ext.kotlinVersion"
```
As "[dependency on the standard library (stdlib) is added automatically to each source set](https://kotlinlang.org/docs/gradle.html#dependency-on-the-standard-library)" we shouldn't need this or any other `stdlib` as direct dependency. WDYT about removing this and rest of them in other projects?

oguzkocer replied:
That's a change I've been wanting to make for a while now, but I strongly prefer making these changes in a standalone PRs especially in WPAndroid where there are quite a few modules. That's unless that change becomes a requirement for this PR.

wzieba commented on build.gradle:
```diff
@@ -42,39 +49,18 @@ ext {
-            }
+    configurations.all {
+        resolutionStrategy {
+            // Don't allow Gradle to resolve the dependency to any other value
+            force "org.jetbrains.kotlin:kotlin-stdlib:$gradle.ext.kotlinVersion"
+            force "org.jetbrains.kotlin:kotlin-stdlib-common:$gradle.ext.kotlinVersion"
+            force "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$gradle.ext.kotlinVersion"
```
@hichamboushaba stumbled recently on similar issue in Woo: https://github.com/woocommerce/woocommerce-android/pull/5521

As he proposed similar solution, we agreed that it's not the best one as we only force version of Kotlin's stdlib, not the whole Kotlin Gradle Plugin. This means that those versions can quickly run out of sync and we believe it might cause problems - probably there's a reason why Kotlin synced those versions out of the box.

Unfortunately, we didn't find any way of forcing specific version of Kotlin Gradle Plugin and our bump is forced by other plugin (`safeargs` in this case).

In conclusion, I believe it's best to leave Kotlin resolution as it is and accept that external plugins can influence version of Kotlin we use. I also try always to believe that there was a reason why external plugin bumped Kotlin/Kotlin Gradle Plugin - it might not work correctly if we force standard library to lower version. 

We eventually always might decide to not just bump some external plugin if we decide that Kotlin bump is dangerous. What we though gain is robust and firm build configuration, while forcing specific Kotlin stdlib version against its plugin's version seems like action against Kotlin build system design.

What are your thoughts @hichamboushaba @oguzkocer ?

hichamboushaba replied:
Thanks for the ping @wzieba, as we discussed before, I agree we should keep the same version that the kotlin's plugin uses for the runtime dependencies as well.

> I believe it's best to leave Kotlin resolution as it is and accept that external plugins can influence version of Kotlin we use

I'm afraid this won't be very safe, kotlin's updates need proper testing to confirm there are no hidden regressions in different Android APIs, and if we accept updates coming from external plugins, they will go unnoticed, unless there is a compile-time regression, so I prefer to hold on updating the external dependency than updating kotlin's version.

I know we didn't find a way to enforce the plugin version, but I think we can at least make the build fail when a different version than what we expect is resolved, we can probably add some [constraints](https://docs.gradle.org/current/userguide/dependency_constraints.html#sec:adding-constraints-transitive-deps) using a `strictly` version, this would make the build fail if a transitive dependency caused an update implicitly, the code would be something like this:

```
    constraints {
        for (module in ["stdlib", "stdlib-jdk7", "stdlib-jdk8"]) {
            implementation("org.jetbrains.kotlin:kotlin-$module") {
                version {
                    strictly gradle.ext.kotlinVersion
                }
            }
        }
    }
```

WDYT about this? I tested this with my navigation's update PR, and this constraint causes the build to fail, while it succeeds when I downgrade the `Navigation` version

wzieba replied:
Sorry for long reply

> I prefer to hold on updating the external dependency than updating kotlin's version.

Sure, my phrasing was not the best but what I meant is to accept the fact of this behaviour and not force our own standard library.

I like the way of handling you proposed. I didn't use constraints but as you say if:

> we can at least make the build fail when a different version than what we expect is resolved

it would be awesome!

oguzkocer replied:
@wzieba I understand where you're coming from, but in practice I don't think it makes sense to blindly accept a transitive dependency version. There might be a case where it's better to update our project and there might be a case where we have to force our own version. Your initial argument doesn't answer recommend a way to address the failure as far as I can tell, so I couldn't compare the trade-offs very well.

> Unfortunately, we didn't find any way of forcing specific version of Kotlin Gradle Plugin and our bump is forced by other plugin (safeargs in this case).

@wzieba @hichamboushaba I think it's possible to force a specific version, let me try it out tomorrow and get back to you.

> I know we didn't find a way to enforce the plugin version, but I think we can at least make the build fail when a different version than what we expect is resolved, we can probably add some constraints using a strictly version, this would make the build fail if a transitive dependency caused an update implicitly, the code would be something like this:

As far as I can tell, this is just a different syntax that'll result in the same thing because `strictly` will also force the same resolution. Having said that, this seems like the better syntax. If we have to force the version, and assuming the constraints get us to the solution we want, I like using this. Thank you for sharing @hichamboushaba.

---

@wzieba @hichamboushaba The plugin version being possibly different was a very important point, thank you for bringing it up, I unfortunately missed this idea in my initial implementation. I think it's possible to add a constraint to the plugin version resolution and once that's done for Kotlin plugin, it might be possible to avoid the extra step by removing the direct dependencies as suggested by @wzieba [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15314#discussion_r775477847). I didn't want to make that change as part of this PR [as I explained in my reply](https://github.com/wordpress-mobile/WordPress-Android/pull/15314#discussion_r778482297), but I definitely don't mind doing it if there is an extra benefit to it.

Fwiw, I think the issue we are dealing is potentially a bug and if we can get away without using these extra constraints, we should remove it. (possibly after upgrading to Kotlin `1.6`) I still think it's a good idea to have more control over how our dependencies are resolved, but I'd prefer a more general solution than just forcing one dependency to be resolved in a certain way.

hichamboushaba replied:
> As far as I can tell, this is just a different syntax that'll result in the same thing because strictly will also force the same resolution. Having said that, this seems like the better syntax. If we have to force the version, and assuming the constraints get us to the solution we want, I like using this.

I'm not sure it's the same thing (but I may be wrong, my knowledge of gradle is very limited 🙂), the `resolutionStrategy` will force the dependency, so the build won't fail, even if a library is bringing a dependency with a higher version, where the `constraints` block will let Gradle do its work for resolving the version it needs, and it will then check it against the stated `constraints`, and the build will fail if they can't be respected.

>  I think it's possible to force a specific version, let me try it out tomorrow and get back to you.

This would be great if possible, thanks.

oguzkocer replied:
@hichamboushaba @wzieba Just wanted to give you a heads up that it'll be a while before I can get back to this PR. I'm going to convert it back to draft and will ping you again when it's ready. Thank you for your feedback so far 🙇‍♂️

oguzkocer replied:
I was able to drop the force resolution rule since we updated Kotlin in `trunk`.

wzieba commented on settings.gradle:
```diff
@@ -1,5 +1,52 @@
+            }
+            if (requested.id.id == "androidx.navigation.safeargs.kotlin") {
+                useModule("androidx.navigation:navigation-safe-args-gradle-plugin:$gradle.ext.navComponentVersion")
+            }
```
I can't say which exactly (no mention in [release notes](https://developer.android.com/jetpack/androidx/releases/navigation)), but after some version (e.g. `2.4.0-beta02`), `safeargs` plugin seems to support plugin marker artifacts and there's no need to define custom `resolutionStrategy`

It might be valuable to add some comment on that to fix this after bumping `navComponentVersion` .

oguzkocer replied:
Addressed in 79fb6cf04c45ec89717ea0aa279c574ad924327e by updating `navComponentVersion` and dropping the resolution rule.

P.S: I'll open a separate PR for the `navComponentVersion` update before marking this PR ready for review, so it won't actually be a part of this PR,, but I wanted to reply to these comments while the changes are fresh in my mind.

wzieba commented on settings.gradle:
```diff
@@ -1,5 +1,52 @@
+            }
+            if (requested.id.id == "com.google.gms.google-services") {
+                useModule("com.google.gms:google-services:4.3.8")
+            }
```
I think that `google-services` doesn't seem to need custom `resolutionStrategy` - at least it works in Woo without problems. WDYT about removing this?

oguzkocer replied:
Thanks for catching this @wzieba! Fixed in 8329b6f5948c4c69b3282b6716d078a640e9beb5.

wzieba commented on build.gradle:
```diff
@@ -1,3 +1,10 @@
+plugins {
+    id "com.automattic.android.fetchstyle"
+    id "io.gitlab.arturbosch.detekt"
+    id "androidx.navigation.safeargs.kotlin" apply false
+    id "com.android.library" apply false
```
Just FYI: last week I had an attempt of fixing this by moving all Android-dependant plugins away from root project but eventually this idea failed because of how Kotlin Gradle Plugin works with Gradle API. Apparently, it seems to be not-fixable. See https://youtrack.jetbrains.com/issue/KT-31643 for details and my attempt: https://github.com/woocommerce/woocommerce-android/pull/5523

wzieba commented on build.gradle:
```diff
@@ -1,3 +1,10 @@
+    id "com.automattic.android.fetchstyle"
+    id "io.gitlab.arturbosch.detekt"
+    id "androidx.navigation.safeargs.kotlin" apply false
+    id "com.android.library" apply false
```
There's `Warning: Please apply google-services plugin at the bottom of the build file.` warning in build log, the same as you suggested in Woo PR (https://github.com/woocommerce/woocommerce-android/pull/5414#discussion_r764338660) I'm suggesting the same 🙂:

```suggestion
    id "com.android.library" apply false
    id 'com.google.gms.google-services' apply false
```

oguzkocer replied:
Addressed in e3c97b8acf216a6c9ea47b4eda3a3d1d45ca3ac1.

oguzkocer committed:
Merge remote-tracking branch 'origin/trunk' into plugin-dsl-2

oguzkocer committed:
Fixes incorrect merge conflict resolution in editor build.gradle

oguzkocer committed:
Remove force resolutionStrategy for kotlin version

oguzkocer committed:
Use plugin DSL for com.google.gms.google-services

oguzkocer committed:
Use plugin DSL for dagger.hilt

oguzkocer committed:
Update navComponentVersion to 2.4.2

oguzkocer committed:
Add 'google-services' plugin to build.gradle without applying

Co-authored-by: Wojciech Zięba <wojtek.zieba@automattic.com>

ParaskP7 commented on settings.gradle:
```diff
@@ -1,3 +1,50 @@
+pluginManagement {
+    gradle.ext.kotlinVersion = '1.6.10'
+    gradle.ext.agpVersion = '4.2.2'
```
Warning (⚠️): @oguzkocer FYI and due to the merge conflicts after updating `trunk`, this dependency wasn't updated to `7.1.1`.

oguzkocer replied:
Nice catch! Fixed in d573663e44a753bd7d06ead2f4be9421f3b41d8a.

ParaskP7 replied:
Great, thanks for the update! 🥇

ParaskP7 commented on libs/editor/WordPressEditor/build.gradle:
```diff
@@ -71,7 +71,7 @@ dependencies {
     implementation "org.wordpress:utils:$wordPressUtilsVersion"
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlinVersion"
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$gradle.ext.kotlinVersion"
     api ("$gradle.ext.aztecAndroidAztecPath") {
```
Question (❓): @oguzkocer I've seen the `aztecVersion` being added and removed as I was traversing the commits in this PR. I am not sure what would be the best, to have that or keep it as is, which is without. Maybe, having the `aztecVersion` around, just like it is done for the `aztecAndroidWordPressShortcodesPath` and `aztecAndroidWordPressCommentsPath` is best, even if only for consistency purposes. Wdyt?

`api ("$gradle.ext.aztecAndroidAztecPath:$aztecVersion") { ... }`

oguzkocer replied:
I think it's a matter of personal preference. Gradle dependency resolution will work the same way for both syntax.

I don't think we should introduce a change within this PR, but if you'd like to open a PR making these consistent across all build files, I am all for it.

ParaskP7 replied:
Great, thanks for the explanation Oguz, I mentioned that just in case, I am ok with either making this change here or as a separate PR, which could maybe improve on all other such inconsistencies. All good from my side and noted! 👍

oguzkocer committed:
Update AGP version to 7.1.1

oguzkocer committed:
Merge remote-tracking branch 'origin/trunk' into plugin-dsl

oguzkocer committed:
Add kotlin parcelize plugin to root build.gradle without applying

Without this change `:libs:editor:WordPressEditor` uses `1.6.0` version of `org.jetbrains.kotlin:kotlin-parcelize-runtime` instead of `1.6.10` which is our current `kotlinVersion`.

oguzkocer commented:
@ParaskP7 This PR is now ready for review. @hichamboushaba @wzieba You're also welcome to review it again if you'd like!

To summarize the recent changes:
* I merged #16632 to `trunk` and then merged `trunk` into this branch which addressed the lint issue we had.
* After the merge commit 4fd566ec4d1f4ac4cf75c5defb7f34980c93836b; [`Dependency Tree Diff` CircleCI job](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Android/31697/workflows/35c3f85a-6349-4ed8-b94a-c5272ff5fad5/jobs/124631) reported that `org.jetbrains.kotlin:kotlin-parcelize-runtime:1.6.0` was added to `:libs:editor:WordPressEditor` instead of `1.6.10` - our current `kotlinVersion`. This I believe is due to differences between how plugin DSL and `buildscript` are interpreted by Gradle. Specifically, in plugin DSL, I find that in some cases a plugin needs to be declared in both `settings.gradle` & root `build.gradle` for it to load correctly to be used by submodules. I tried this in 7d0c5b8fc15fb0eeb2619dd1f73c219b1bbe42b2 which fixed the issue.
* Note that while working on 7d0c5b8fc15fb0eeb2619dd1f73c219b1bbe42b2, I accidentally reset my `git config` as I ran `./tools/dependency-tree-diff/dependency-tree-diff.sh` script without cleaning up the CircleCI related commands. I ended up having to force push to fix the author of the latest commit without making any code changes. I am against force pushing after a PR is reviewed, but in this case this was the only way I know to restore the author information for the commit and since it didn't contain any code changes, I thought it to be a reasonable exception.

ParaskP7 commented:
> @ParaskP7 This PR is now ready for review.

Woohoo, great, thank you so much for working on this @oguzkocer , I'll be reviewing that today! 💯

ParaskP7 commented on libs/image-editor/ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -271,14 +271,16 @@ class PreviewImageFragment : Fragment(R.layout.preview_image_fragment) {
-                    ),
-                    navOptions
-            )
+            outputFileExtension?.let {
```
Warning (⚠️): This is most probably a merge conflict mistake that needs to be reverted, we already reverted this change in the accompanying navigation component PR [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16632/commits/51af542de4a64d6a26681cf5c5b2e2131f94e676), isn't that right?

oguzkocer replied:
Good catch! Unfortunately `git` decided to resolve the conflict itself and did a poor job of it. I did a pass through afterwards, but missed this, so thank you for catching it! 🙇‍♂️ 

Fixed in f34d4ce4ded43383ac4505812cc21ce49bb35d6d.

ParaskP7 replied:
Coolio, thank you so much for this @oguzkocer , I'll update my review to 100% ✅ ! 🥇 

> Unfortunately git decided to resolve the conflict itself and did a poor job of it.

😅 🤷 🤣

ParaskP7 reviewed (changes requested):
👋 @oguzkocer !

I reviewed and tested this PR again, this time even more throroughtly with my general smoke testing, everything looks good and I didn't encounter any hiccups.

Needless to say, this work is VERY VALUABLE, will help us a lot going forward and THANK YOU for wrangling on it! ❤️ 🥇 💯 

1. Warning (⚠️): I didn't approve at this point just because of this single merge conflict related warning comment I raised. Other than that it is a ✅ from my side!
2. Suggestion (💡): I would also recommend for us to trigger the UI tests before merging this to `trunk`, just so we reach an additional level of confidence. PS: I generally recommend doing that when we have PR such as this or with dependency update PRs, but granted, I sometimes forget to suggest that (like I forgot [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16632)) or I forget triggering that on my own PRs. 😊

PS: @hichamboushaba @wzieba do let us know if we are missing something and thank you so much for helping on this since day one! 🙇

oguzkocer committed:
Revert changes due to git auto merge-conflict resolution
